### ISSUE TITLE
Rationalizing Symbol.species

### ISSUE THREAD

**Issue by bakkot** [tc39/ecma262#1313]
In light of the committee's decision to reject #1289, and the upcoming [set methods](https://github.com/tc39/proposal-set-methods) proposal (which would be one of the first usages of `Symbol.species` in the spec which worked for methods invoked on non-subclasses) it would be valuable for the committee to decide what our rationale for `Symbol.species` is. In particular:

- What non-normative text should describe it?
- What criteria should we use to decide when `Symbol.species` is used, going forward?
- Maybe other questions?

I've suggested some options below. Please react with 👍 and 👎 to indicate your preferences, and suggest other alternatives if you have any.

cc @wycats @gsathya

**Comment by bakkot** [comment]
One possible alternative: decide Arrays are just weird (as they are anyway, for cross-realm web compat reasons), and we should continue to make new generic methods use Symbol.species even when invoked on things which are not instances of the class on which the original method was defined.

**Comment by bakkot** [comment]
One possible alternative: decide that `Symbol.species` is specifically intended for use with subclasses, and introduce a brand check before doing the `Symbol.species` lookup in new methods. For example, the set methods proposal could introduce a `SetSpeciesCreate` which performed a brand check to see if the methods' callee is a `Set`, and if not just produce a new `Set` rather than looking up the constructor and `Symbol.species`.

**Comment by bakkot** [comment]
One possible alternative: decide that `Symbol.species` is specifically intended for use with subclasses, but we're not going to introduce our own brand checks. Make a non-normative change to the description of `Symbol.species` making this clear, but don't add brand checks to new methods when they're not otherwise necessary.

We might consider using [MDN's description](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/species#Description) for `Symbol.species` as a starting point:

> The species accessor property allows subclasses to override the default constructor for objects.

**Comment by bakkot** [comment]
One possible alternative: change the behavior of `Array.prototype.{concat, filter, map, slice, splice}` to not access `Symbol.species` even on `Array` subclasses, and instead always create a new `Array`. This makes subclassing `Array` less useful, since you'd likely want to introduce your own versions of these methods, and would be breaking if people are subclassing Array and relying on this behavior.

We could rationalize this by saying that those methods need to have their own magic for cross-realm stuff for web-compat reasons, which is true.

**Comment by domenic** [comment]
One possible alternative: consider `Symbol.species` a legacy mistake, and not use it on any newly-introduced methods. Consider removing it from any existing methods where doing so would be web-compatible.

**Comment by bathos** [comment]
An example of a case where @@species was more useful than the alternative for consideration:

I once had a Promise subclass where it was important to spawn ordinary promises (or else they’d have cycled). I could have done this by overriding `then` and `catch`. But if I had done that instead of using `@@species`, this class would have had its API broken by the introduction of `finally`. I believe one important utility of `@@species` is that it makes subclasses more likely to be forward compatible.

**Comment by domenic** [comment]
> But if I had done that instead of using @@species, this class would have had its API broken by the introduction of finally

Nope. `catch` and `finally` both delegate to `then`, so overriding `then` was sufficient.

**Comment by getify** [comment]
I too have a [Promise subclass](https://gist.github.com/getify/1173cac45d15fc4ff0a880f32fd598ab) that benefits from Symbol.species. I hope it doesn't go away.

**Comment by bathos** [comment]
@domenic It delegates to `then`, yeah — so why does the `finally` algorithm call SpeciesConstructor? Sorry, tangential question, but I’m pretty curious now.

I think the point about reducing forward compat risks likely stands even if the example I gave was poor. Do you think that is not the case? Or that it is, but its value doesn’t outweight the cost of @@species?

**Comment by domenic** [comment]
`finally` delegates to SpeciesConstructor because of legacy design mistakes carried forward in the name of consistency.

**Comment by bathos** [comment]
I gather you feel strongly about this haha! I’d like to understand why you concluded that it was a legacy design mistake though.

(I read through #1289, but it didn’t quite fill in the gap — anybody have a link to prior discussion that would clarify?)

**Comment by littledan** [comment]
OK, it sounds like there are some uses of Promise subclassing, but it's not clear whether species is necessary to achieve the desired functionality or whether overriding other methods would be sufficient.

Are there known cases for other classes creating an instance of the subclass from methods through species, e.g., TypedArray, Array or RegExp?

**Comment by claudepache** [comment]
We can certainly justify the brand-check in ArraySpeciesCreate() as a way to support the legacy way to apply array methods to non-arrays, casting the result into an array in the process. But I don’t think it is a good reason to extend  such brand-check to new stuff; we should rather encourage explicit casting functions à la `Array.from()`.

Another things that bugs me, is that, when [SpeciesConstructor()](https://tc39.github.io/ecma262/#sec-speciesconstructor) or similar is invoked, a _default constructor_ is provided. I think it is a design error (notwithstanding legacy stuff like arrays), because you can’t really guess what the intended constructor is. For example:

```js
Array.prototype.slice.call(nodeList)
```
will cast a NodeList instance into an Array object, but

```js
class BetterArray extends Array { }
BetterArray.prototype.slice.call(nodeList)
```
will confusingly *not* cast a NodeList instance into a BetterArray object.

More generally, any method whose algorithm contains an explicit reference to its class/constructor is expected to behave surprisingly when inherited by a subclass.

**Comment by bakkot** [comment]
See further discussion of this topic in https://github.com/tc39/proposal-rm-builtin-subclassing.

**Comment by getify** [comment]
I can see and support the arguments for removing Type III and Type IV subclassing. But removal of Type II subclassing, IMO, significantly neuters the idea of subclassing altogether. I would expect to abandon any efforts to use subclassing if Type II were removed.

The fact that the decision criteria is largely based around security and web-compat is relevant, but also (IMO) short-sighted. Not enough weight is given to coherence of a feature and typical developer intuition -- which only gets a passing glance of a mention in that linked explainer. The "Exit" section at the end concludes that Type II would only be abandoned if it proved web-incompatible to remove, leaving no room for preserving it based on its own merits (despite its cost).

I concur with the explainer that Type II is strongly aligned with developer intuition, and I also (as mentioned earlier in this thread) have built code that relies on it. Thus, its merits are that it's what a lot of developers expect, and it's an assumption developers use when they build code. Whether it's web-compatible to remove or not, it should at least have a vigorous debate and defense rather than concluding that it's already a given that it should be removed if possible.

Discarding Type II (taking that away) after 6 years because it tends to make implementations hard to secure and/or optimize is a much more drastic ask than removing the Type III and Type IV variants, which (by comparison, in my estimation) are far less common or intuitively expected.

**Comment by getify** [comment]
As a sort of illustrative aside, I have adopted a number of programming patterns (built on libraries I've authored) in recent years that are heavily focused on FP techniques, especially monads of late. I'm well aware that despite any best efforts to optimize the implementations of these libraries, usage of them comes at a performance cost. The low-level non-FP non-monadic way of doing a task is almost always going to beat, and in some cases, significantly beat, these more recently adopted approaches.

The fact that in JS you have the option to a variety of paths on the style/performance spectrum, according to the needs of any given parts of a program is, IMO, one of JS's greatest strengths. But not all code paths are critical code paths. In fact, most aren't. So I choose to continue use coding styles that I know are less efficient because raw performance, in most of my code, is not the primary concern.

Following coding conventions and standards that are deemed to be more "proper" (for varying definitions of that word), hoping that it avoids bugs and creates a more expressive and readable and maintainable program -- these are overriding concerns compared to a few dozen microseconds longer execution time. Only if I see a truly hot code path being significantly harmed would I consider optimizing away these patterns I've come to trust and rely on.

My point? Developer intuition and expected patterns (even as inspired by experience from other languages) should be favored, or at least given more weight than it seems to be getting, for the majority of lines in a program. Just because Type II subclassing may make code a bit slower doesn't mean it's in any way a bad choice for a majority of code. It might very well need to be avoided in some limited hot paths, but that should be up to the author, not the language steering committee.
